mlserver/batch_processing.py

```diff
@@ -81,87 +82,155 @@ def get_headers(request_id: str) -> Dict[str, str]:
-        )
-        inputs.append(new_input)
+@dataclass
+class TritonRequest:
+    id: str
+    inputs: List[httpclient.InferInput]
+    outputs: List[httpclient.InferRequestedOutput] = None
+
+    @classmethod
+    def from_inference_request(
+        cls, inference_request: InferenceRequest, binary_data: bool
+    ) -> Tuple[str, List[httpclient.InferInput], List[httpclient.InferRequestedOutput]]:
+        inputs = []
+
+        for request_input in inference_request.inputs or []:
+            new_input = httpclient.InferInput(
+                request_input.name, request_input.shape, request_input.datatype
+            )
+            new_input.set_data_from_numpy(
+                NumpyCodec.decode_input(request_input),
+                binary_data=binary_data,
+            )
+            inputs.append(new_input)
 
-    outputs = []
-    for request_output in inference_request.outputs or []:
-        new_output = httpclient.InferRequestedOutput(
-            request_output.name, binary_data=binary_data
-        )
-        outputs.append(new_output)
+        outputs = []
+        for request_output in inference_request.outputs or []:
+            new_output = httpclient.InferRequestedOutput(
+                request_output.name, binary_data=binary_data
+            )
+            outputs.append(new_output)
 
-    request_id = (
-        generate_uuid() if inference_request.id is None else inference_request.id
-    )
-    return request_id, inputs, outputs
+        return TritonRequest(inference_request.id, inputs, outputs)
 
 
-def serialize_triton_infer_result(output_item: BatchOutputItem) -> bytes:
-    response = output_item.item.get_response()
-    if "parameters" not in response:
-        response["parameters"] = {}
-    response["parameters"]["batch_index"] = output_item.index
-    for (n, output) in enumerate(response["outputs"]):
-        data = output_item.item.as_numpy(output["name"])
-        response["outputs"][n]["data"] = data.flatten().tolist()
+def infer_result_to_infer_response(item: httpclient.InferResult) -> InferenceResponse:
+    response = item.get_response()
 
-        # Removing "binary_data_size" parameters as now this is 1-D list
-        if (
-            output.get("parameters") is not None
-            and "binary_data_size" in output["parameters"]
-        ):
-            del response["outputs"][n]["parameters"]["binary_data_size"]
+    outputs = []
+    for output_item in response["outputs"]:
+        output_name = output_item["name"]
+        output = NumpyCodec.encode_output(output_name, item.as_numpy(output_name))
+        output_parameters = output_item["parameters"]
+        if output_parameters is not None:
+            # Create copy in case this code will ever be used in another context as
+            # we are modifying the original parameters by removing unwanted key
+            output_parameters = output_parameters.copy()
+            if "binary_data_size" in output_parameters:
```

Comment:
No harm in leaving extra parameters here BTW - although I can see how it would look weird to have many random ones on the response.

Reply:
They make sense only in the initial response from Triton when this is really encoded as binary data. 
When we call `item.as_numpy(name)` we decode that so the parameters is not longer relevant nor valid.